pkg/apiserver/registry/system/bundle/rest.go

```diff
@@ -0,0 +1,367 @@
+// which has the same name of the mode.
+func (r *bundleREST) Get(_ context.Context, name string, _ *metav1.GetOptions) (runtime.Object, error) {
+	r.statusLocker.RLock()
+	defer r.statusLocker.RUnlock()
+	if r.cache.Name != name {
+		return nil, errors.NewNotFound(systemv1beta1.Resource("bundle"), name)
+	}
+	return r.cache, nil
+}
+
+// Delete can remove the current finished bundle or cancel a running bundle
+// collecting. It only allows to query the resource which has the same name of
+// the mode.
+func (r *bundleREST) Delete(_ context.Context, name string, _ rest.ValidateObjectFunc, _ *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	if name != r.mode {
+		return nil, false, errors.NewNotFound(systemv1beta1.Resource("bundle"), name)
+	}
+	r.statusLocker.Lock()
+	defer r.statusLocker.Unlock()
+	if r.cancelFunc != nil {
+		r.cancelFunc()
+	}
+	r.cache = &systemv1beta1.Bundle{
+		ObjectMeta: metav1.ObjectMeta{Name: r.mode},
+		Status:     systemv1beta1.BundleStatusNone,
+	}
+	return nil, true, nil
+}
+
+func (r *bundleREST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *bundleREST) collect(ctx context.Context, dumpers ...func(string) error) (*systemv1beta1.Bundle, error) {
+	basedir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+	if err != nil {
+		return nil, fmt.Errorf("error when creating tempdir: %w", err)
+	}
+	defer defaultFS.RemoveAll(basedir)
+	for _, dumper := range dumpers {
+		if err := dumper(basedir); err != nil {
+			return nil, err
+		}
+	}
+	outputFile, err := defaultFS.Create(filepath.Join(afero.GetTempDir(defaultFS, ""), fmt.Sprintf("bundle_%d.tar.gz", rand.Int())))
+	if err != nil {
+		return nil, fmt.Errorf("error when creating output tarfile: %w", err)
+	}
+	defer outputFile.Close()
+	hashSum, err := packDir(basedir, outputFile)
+	if err != nil {
+		return nil, fmt.Errorf("error when packaing bundle: %w", err)
+	}
+
+	select {
+	case <-ctx.Done():
+		_ = defaultFS.Remove(outputFile.Name())
+		return nil, fmt.Errorf("collecting is canceled")
+	default:
+	}
+	stat, err := outputFile.Stat()
+	var fileSize int64
+	if err == nil {
+		fileSize = stat.Size()
+	}
+	creationTime := metav1.Now()
+	deletionTime := metav1.NewTime(creationTime.Add(bundleExpireDuration))
+	return &systemv1beta1.Bundle{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              r.mode,
+			CreationTimestamp: creationTime,
+			DeletionTimestamp: &deletionTime,
+		},
+		Status:   systemv1beta1.BundleStatusCollected,
+		Sum:      fmt.Sprintf("%x", hashSum),
+		Size:     uint32(fileSize),
+		FilePath: outputFile.Name(),
+	}, nil
+}
+
+func (r *bundleREST) collectAgent(ctx context.Context) (*systemv1beta1.Bundle, error) {
+	agentDumper := agentutil.NewDumper(defaultFS, defaultExecutor, ofctlClientFactory)
+	generalDumper := util.NewDumper(defaultFS, defaultExecutor)
+	return r.collect(
+		ctx,
+		generalDumper.DumpAgentLog,
+		generalDumper.DumpOVSLog,
+		agentDumper.DumpIPToolInfo,
+		agentDumper.DumpIPTables,
```

Comment:
These two are for Linux only. Again not sure if we want to expose Linux or Windows agent logic here.

Do you think we should just encapsulate all these to three funcs like: DumpAgentBundle, DumpControllerBundle, DumpNetworkPolicyBundle? The first two are in agent/ and controller/ respectively, and the last one is shared (and can be called by the first two).